### PR TITLE
Implement CompactFiles API

### PR DESCRIPTION
**Pull request by huachaohuang**
CompactFiles is faster than CompactRange when we want to do full compaction because CompactFiles can take good use of multiple threads. But CompactFiles can only be used when auto compaction is disabled.

### CODE REVIEW

File: src/util/rocksdb/mod.rs [hunk 1]
```diff
@@ -341,6 +341,33 @@ pub fn compact_range(
     db.compact_range_cf_opt(handle, &compact_opts, start_key, end_key);
 }
 
+pub fn compact_files(db: &DB, cf: &CFHandle, output_level: u32) -> Result<(), String> {
```
Comment (siddontang) [inline src/util/rocksdb/mod.rs]: do we need to move this to the RocksDB repo?

**Comment by siddontang** [comment]
> But CompactFiles can only be used when auto compaction is disabled.

If we enable auto-compaction, can we disable it explicitly when compact file then enable it?

File: src/util/rocksdb/mod.rs (hunk 1 above)
Comment (huachaohuang) [reply src/util/rocksdb/mod.rs]: I don't think so, this is just a specific use case for us, like other RocksDB related code here.

**Comment by huachaohuang** [comment]
It doesn't seem like we can disable auto-compaction on the fly.
I only use this API when I know exactly that auto-compaction is disabled (otherwise it may return an error), we can improve it later if necessary.

**Comment by siddontang** [comment]
see https://github.com/facebook/rocksdb/blob/master/include/rocksdb/options.h#L258

Seem that we can dynamically enable/disable the auto-compaction now.

**Comment by huachaohuang** [comment]
Although we can enable/disable auto-compaction online, if there are some ongoing compactions when we compact files, it may still return errors. In addition, in my use case, I don't want to enable auto-compaction after compacting files, so I think whether to enable/disable auto-compaction is use case related. We can support a more general API to change options dynamically when necessary.

File: src/util/rocksdb/mod.rs [hunk 2]
```diff
@@ -341,6 +341,33 @@ pub fn compact_range(
...
+    let compressions = cf_opts.get_compression_per_level();
+
+    let mut opts = CompactionOptions::new();
+    if let Some(compression) = compressions.get(output_level as usize) {
+        opts.set_compression(*compression);
+    }
+    opts.set_output_file_size_limit(cf_opts.get_target_file_size_base() as usize);
+
+    let mut input_files = Vec::new();
+    let cf_meta = db.get_column_family_meta_data(cf);
```
Comment (zhangjinpeng1987) [inline src/util/rocksdb/mod.rs]: We should disable auto compaction and confirm there is no compaction is running before get all files.

**Comment by siddontang** [comment]
@huachaohuang 

Now I guess we should support changing configuration dynamically at first.

**Comment by huachaohuang** [comment]
I think `CompactFiles` has too many limitations and not practical for the general use case, even if we change options dynamically, so I close this now. BTW, there is an ongoing PR to speed up `CompactRange` https://github.com/facebook/rocksdb/pull/3549.

**Pull request closed without merging**